File: jsrc/machine/SxxMachine/Trail.java
package com.googlecode.prolog_cafe.lang;

import java.util.ArrayDeque;
import java.util.Deque;

/**
 * Trail stack.<br>
 * The class <code>Trail</code> represents a trail stack.<br>
 * Entries pushed to this trail stack must implement
 * the <code>Undoable</code> interface.
 * @see Undoable
 * @author Mutsunori Banbara (banbara@kobe-u.ac.jp)
 * @author Naoyuki Tamura (tamura@kobe-u.ac.jp)
 * @version 1.0
 */
public final class Trail {
	/** list of already filled buffers */
	private final Deque<Undoable[]> buffersList;

	private final int pageSize;
	/** An array of <code>Undoable</code> entries. */
	private Undoable[] buffer;

	/** the top index of this <code>Trail</code>. */
	private int top;

	private int base;

	/** Current timestamp of the top of {@link ChoicePointStack}. */
	long timeStamp;

	/** Constructs a new trail stack. */
	public Trail() {
		this(2048);
	}

	/** Constructs a new trail stack with the given page size. */
	public Trail(int pageSize) {
		this.pageSize = pageSize;
		buffersList = new ArrayDeque<Undoable[]>(1000);
		buffer = new Undoable[pageSize];
		top = -1;
		base = 0;
	}

	/** Constructs a new trail stack with the given page size and initial number of pages. */
	public Trail(int pageSize, int pages) {
		this.pageSize = pageSize;
		buffersList = new ArrayDeque<Undoable[]>(pages);
		buffer = new Undoable[pageSize];
		top = -1;
		base = 0;
	}

	/** Discards all entries. */
	public void init() { deleteAll(); }

	/** Pushs an entry to this <code>Trail</code>. */
	public void push(Undoable t) {
		if (++top >= pageSize) {
			buffersList.addLast(buffer);
			buffer = new Undoable[pageSize];
			top = 0;
			base+=pageSize;
		}
		buffer[top] = t;
	}

	/** Pops an entry from this <code>Trail</code>. */
	public Undoable pop() {
		Undoable t = buffer[top];
		buffer[top--] = null;
		if (top<0 && base>0){
			buffer = buffersList.pollLast();
			top = buffer.length-1;
			base -= pageSize;
		}
		return t;
	}

	/** Discards all entries. */
	private void deleteAll() {
		buffersList.clear();
		buffer = new Undoable[buffer.length];
		top = -1;
		base = 0;
	}

	/** Tests if this stack has no entry. */
	public boolean empty() {
		return base==0 && top == -1;
	}

	/** Current allocation of the trail storage array.  */
	public int max() { return base+pageSize; }

	/** Returns the value of <code>top</code>.
	 * @see #top
	 */
	public int top() { return base + top; }

	/** Unwinds all entries after the value of <code>i</code>. */
	public void unwind(int i) {
		while (base + top > i) {
			Undoable t = buffer[top];
			buffer[top--] = null;
			if (top<0 && base>0){
				buffer = buffersList.pollLast();
				top = pageSize-1;
				base -= pageSize;
			}
			t.undo();
		}
	}

//    /** Shows the contents of this <code>Trail</code>. */
//    public void show() {
//	if (empty()) {
//	    System.out.println("{trail stack is empty!}");
//	    return;
//	}
//	for (int i=0; i<=top; i++) {
//	    System.out.print("trail[" + i + "]: ");
//	    System.out.println(buffer[i]);
//	}
//    }
}

File: src/machine/SxxMachine/Trail.cpp
using namespace std;

#include "Trail.h"
#include "Undoable.h"
#include "Prolog.h"

namespace SxxMachine
{

	Trail::Trail(Prolog* m) : Trail(m,2048)
	{
	}

	Trail::Trail(Prolog* m, const int& pageSize) : Trail(m, pageSize, 1000)
	{
	}

	Trail::Trail(Prolog* m, const int& pageSize, const int& pages) : buffersList(new ArrayDeque<std::vector<Undoable*> >(pages)), pageSize(pageSize), mach(m)
	{
		this->buffer = std::vector<Undoable*>(pageSize);
		this->top_RenamedTODO = -1;
		this->base = 0;
	}

	void Trail::init()
	{
		deleteAll();
	}

	void Trail::push(Undoable* t)
	{
		if(++this->top_RenamedTODO >= this->pageSize)
		{
			this->buffersList->addLast(this->buffer);
			this->buffer = std::vector<Undoable*>(this->pageSize);
			this->top_RenamedTODO = 0;
			this->base += this->pageSize;
		}
		this->buffer[this->top_RenamedTODO] = t;
	}

	Undoable* Trail::pop()
	{
		Undoable* t = this->buffer[this->top_RenamedTODO];
		this->buffer[this->top_RenamedTODO--] = nullptr;
		if(this->top_RenamedTODO < 0 && this->base > 0)
		{
			this->buffer = this->buffersList->pollLast();
			this->top_RenamedTODO = this->buffer.size() - 1;
			this->base -= this->pageSize;
		}
		return t;
	}

	void Trail::deleteAll()
	{
		this->buffersList->clear();
		this->buffer = std::vector<Undoable*>(this->buffer.size());
		this->top_RenamedTODO = -1;
		this->base = 0;
	}

	bool Trail::empty()
	{
		return this->base == 0 && this->top_RenamedTODO == -1;
	}

	int Trail::max()
	{
		return this->base + this->pageSize;
	}

	int Trail::top()
	{
		return this->base + this->top_RenamedTODO;
	}

	void Trail::unwind(const int& i)
	{
		while(this->base + this->top_RenamedTODO > i)
		{
			Undoable* t = this->buffer[this->top_RenamedTODO];
			this->buffer[this->top_RenamedTODO--] = nullptr;
			if(this->top_RenamedTODO < 0 && this->base > 0)
			{
				this->buffer = this->buffersList->pollLast();
				this->top_RenamedTODO = this->pageSize-1;
				this->base -= this->pageSize;
			}
			t->undo();
		}
	}

	Prolog* Trail::getProlog()
	{
		// TODO Auto-generated method stub
		return this->mach;
	}
}
